src/ucp/core/ucp_am.c

```diff
@@ -558,189 +559,181 @@ ucp_am_find_first_rdesc(ucp_worker_h worker, ucp_ep_ext_proto_t *ep_ext,
+static UCS_F_ALWAYS_INLINE void
+ucp_am_copy_data_fragment(ucp_recv_desc_t *first_rdesc, void *data,
+                          size_t length, size_t offset)
+{
+    memcpy(UCS_PTR_BYTE_OFFSET(first_rdesc + 1, offset), data, length);
+
+    first_rdesc->am_first.remaining -= length;
+}
+
+static UCS_F_ALWAYS_INLINE void
+ucp_am_handle_unfinished(ucp_worker_h worker, ucp_recv_desc_t *first_rdesc,
                          void *data, size_t length, size_t offset)
 {
     uint16_t am_id;
     ucs_status_t status;
-    ucp_am_first_hdr_t *f_hdr;
+    ucp_am_first_hdr_t *first_hdr;
+    void *msg;
     ucp_ep_h reply_ep;
 
-    ucs_assert(f_rdesc->am_first.data != NULL);
-
-    memcpy(UCS_PTR_BYTE_OFFSET(f_rdesc->am_first.data + 1, offset),
-           data, length);
+    ucp_am_copy_data_fragment(first_rdesc, data, length, offset);
 
-    f_rdesc->am_first.left -= length;
-    if (f_rdesc->am_first.left > 0) {
+    if (first_rdesc->am_first.remaining > 0) {
         /* not all fragments arrived yet */
-        return UCS_INPROGRESS;
+        return;
     }
 
-    f_hdr = (ucp_am_first_hdr_t*)(f_rdesc + 1);
-    am_id = f_hdr->super.super.am_id;
+    first_hdr = (ucp_am_first_hdr_t*)(first_rdesc + 1);
+    am_id     = first_hdr->super.super.am_id;
+    msg       = first_hdr + 1;
 
-    if (!ucp_am_recv_check_id(worker, am_id)) {
-        status = UCS_OK;
-    } else {
-        reply_ep = (f_hdr->super.super.flags & UCP_AM_SEND_REPLY)        ?
-                   ucp_worker_get_ep_by_ptr(worker, f_hdr->super.ep_ptr) :
+    if (ucs_likely(ucp_am_recv_check_id(worker, am_id))) {
+        reply_ep = (first_hdr->super.super.flags & UCP_AM_SEND_REPLY)        ?
+                   ucp_worker_get_ep_by_ptr(worker, first_hdr->super.ep_ptr) :
                    NULL;
-        status   = worker->am.cbs[am_id].cb(worker->am.cbs[am_id].context,
-                                            f_rdesc->am_first.data + 1,
-                                            f_hdr->total_size, reply_ep,
+        status   = worker->am.cbs[am_id].cb(worker->am.cbs[am_id].context, msg,
+                                            first_hdr->total_size, reply_ep,
                                             UCP_CB_PARAM_FLAG_DATA);
-    }
-
-    if (status != UCS_INPROGRESS) {
-        ucs_free(f_rdesc->am_first.data); /* user does not need to hold this data */
+    } else {
+        status = UCS_OK;
     }
 
     /* message assembled, remove first fragment descriptor from the list in
      * ep AM extension */
-    ucs_list_del(&f_rdesc->am_first.list);
-
-    return UCS_OK;
-}
+    ucs_list_del(&first_rdesc->am_first.list);
 
-static UCS_F_ALWAYS_INLINE int
-ucp_am_check_rdesc_msg_id(ucp_recv_desc_t *rdesc, uint64_t msg_id)
-{
-    ucp_am_mid_hdr_t *m_hdr = (ucp_am_mid_hdr_t*)(rdesc + 1);
+    if (status != UCS_INPROGRESS) {
+        /* user does not need to hold this data */
+        ucs_free(first_rdesc);
+        return;
+    }
```

Comment:
can we merge this into lines 608/612? for example in line 612 we can do ucs_free unconditionally (maybe "goto out_free_first"). also maybe move ucs_list_del() before line 604